### PR TITLE
Add scale ticket verification to MassID weighing rule

### PR DESCRIPTION
**Pull request by sangalli**
### 🧾 Summary

Introduce shared text and scale ticket extractor libraries and integrate scale ticket verification into the MassID weighing rule to validate event net weight against scale tickets.

### 🧩 Context

**Why this change?** MassID weighing currently validates net weight using document data only, without verifying against the underlying physical scale ticket.  
**Problem it solves:** There is no automated way to detect discrepancies between the MassID weighing event value and the weight recorded in a recycler’s scale ticket, reducing trust in the reported values.  
**Business value:** Adds an auditable, automated verification step that cross-checks MassID weighing data with scale tickets, improving methodology integrity and reducing the risk of misreported weights.

### 🧱 Scope of this PR

**This PR includes:**

- [x] New shared `@carrot-fndn/shared/text-extractor` library wrapping Textract and exposing a text extractor service and provider.
- [x] New shared `@carrot-fndn/shared/scale-ticket-extractor` library with layout-specific parsers, helpers, and tests for extracting weights from scale tickets.
- [x] Updates to bold methodology types and enums to support scale ticket attachments and additional verification configuration on accreditation events.
- [x] Integration of scale ticket verification into the MassID `weighing` rule, including configuration extraction from accreditation, S3 key construction, and rule-level error reporting.
- [x] New and updated tests for the extractor libraries, weighing processor, weighing lambda, and weighing helpers to cover the new verification path and error scenarios.

**Intentionally excluded (future work):**

- Additional scale ticket layouts beyond the initial `layout1`.
- Reuse of the shared text/scale ticket extractors by other rules or methodologies outside MassID weighing.

### 🧪 How to test

1. **Unit tests for shared libraries**
   - Run shared text extractor tests:  
     `pnpm nx test shared-text-extractor`
   - Run shared scale ticket extractor tests:  
     `pnpm nx test shared-scale-ticket-extractor`

2. **Unit/integration tests for MassID weighing rule**
   - Run weighing rule unit tests (processor, helpers, verification):  
     `pnpm nx test methodologies-bold-rule-processors-mass-id-weighing`

3. **Affected tests (optional broader check)**
   - Run only affected tests:  
     `pnpm test:affected`

4. **Manual verification (optional)**
   - Configure an accreditation document with `Required Additional Verifications` including a scale ticket verification entry.
   - Attach a scale ticket file (with the supported `layout1`) to the MassID weighing event using the `Scale Ticket` attachment label.
   - Invoke the weighing rule lambda and confirm:
     - Matching net weight → rule passes with the existing success message.
     - Mismatched net weight or missing/invalid ticket source → rule fails with the new scale ticket–specific error messages.

### 🧠 Considerations for Review

**Key areas to focus on:**

- `libs/shared/text-extractor`: Abstractions over Textract, S3 input types, and how errors are surfaced to callers.
- `libs/shared/scale-ticket-extractor`: Layout-specific parsing logic, tolerance assumptions, and how extracted net weight is computed from ticket fields.
- `libs/methodologies/bold/rule-processors/mass-id/weighing`: 
  - How additional verification configuration is read from accreditation events.
  - How scale ticket verification is wired into the weighing evaluation flow and how failures are mapped to rule output comments.
  - New helper functions and test cases that exercise scale ticket success and failure paths.
- Type and enum changes in bold shared types to ensure they remain backwards compatible for existing rules.

**Potential follow-ups:**

- Support for additional scale ticket layouts and locales.
- Reuse of the extractor libraries in other rules that depend on external document verification (e.g., other methodology events that rely on scale tickets).

---

- [x] **I, the PR author, confirm that this PR works as expected and does not break any service. I also confirm that I applied best practices and improved the codebase quality.**

## Summary by CodeRabbit

* **New Features**
  * Automatic scale‑ticket verification for weighing events with layout‑aware parsing and clear failure messages.
  * New text‑extraction service backed by AWS Textract with runtime type checks.

* **Tests**
  * Extensive unit and e2e tests covering extraction, layout‑1 parsing, verification flows, and edge cases.

* **Other**
  * Consistent attachment S3 key helper, new metadata enum entries, new extractor packages/configs and path aliases, plus lint/jest setup.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: libs/shared/text-extractor/src/textract.service.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      .filter((block) => block.BlockType === 'LINE')
+      .map((block) => block.Text ?? '')
+      .join('\n')
+      .trim();
+
+    if (text.length === 0) {
+      throw new Error('No LINE blocks returned from Textract');
+    }
+
+    return text as TextExtractionResult['rawText'];
```
Comment (AMarcosCastelo) [inline libs/shared/text-extractor/src/textract.service.ts]: Is it valid to add an runtime validation? Maybe an `isNonEmptyString`?

File: libs/shared/scale-ticket-extractor/src/layout-1/scale-ticket-layout-1.helpers.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  );
+
+  if (!isValid(parsed)) {
+    return undefined;
+  }
+
+  return parsed;
+};
+
+const WEIGHT_VALIDATION_TOLERANCE = 1;
```
Comment (AMarcosCastelo) [inline libs/shared/scale-ticket-extractor/src/layout-1/scale-ticket-layout-1.helpers.ts]: What do you think to declare this tolerance in an opened scope and remove from the libs/methodologies/bold/rule-processors/mass-id/weighing/src/weighing.constants.ts?

File: libs/shared/text-extractor/src/text-extractor.types.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import type { NonEmptyString } from '@carrot-fndn/shared/types';
```
Comment (AMarcosCastelo) [inline libs/shared/text-extractor/src/text-extractor.types.ts:1]: I think is valid to rename the file to `text-extractor.types.ts` to keep the pattern.

File: libs/methodologies/bold/rule-processors/mass-id/weighing/src/weighing.processor.ts [hunk 4]
```diff
@@ -171,6 +213,43 @@ export class WeighingProcessor extends RuleDataProcessor {
...
+    }
+
+    const bucket = process.env['DOCUMENT_ATTACHMENT_BUCKET_NAME'];
+
+    if (!bucket) {
+      logger.warn(
+        'DOCUMENT_ATTACHMENT_BUCKET_NAME environment variable is not configured',
+      );
+
+      return undefined;
```
Comment (AMarcosCastelo) [inline libs/methodologies/bold/rule-processors/mass-id/weighing/src/weighing.processor.ts:237]: Do we must return undefined? If the variable is not configured, will the rule skip it?

**Review by AMarcosCastelo — CHANGES_REQUESTED** [review]

File: libs/methodologies/bold/rule-processors/mass-id/weighing/src/weighing.processor.ts (hunk 4 above)
Comment (sangalli) [reply libs/methodologies/bold/rule-processors/mass-id/weighing/src/weighing.processor.ts:237]: This function builds the input to validate the file. When the input is actually used, it checks the value
CleanShot 2025-12-16 at 06 59 18@2x

File: libs/shared/scale-ticket-extractor/src/layout-1/scale-ticket-layout-1.helpers.ts (hunk 2 above)
Comment (sangalli) [reply libs/shared/scale-ticket-extractor/src/layout-1/scale-ticket-layout-1.helpers.ts]: Good point. As the isValid attribute from the ScaleTicketData in not being used, I just removed it for simplicity.

**Review by AMarcosCastelo — APPROVED** [review]
